homeassistant/components/icloud.py

```diff
@@ -0,0 +1,952 @@
+
+                        if strnow < strstart:
+                            ievent = IEvent(self.hass, self, guid,
+                                            TYPE_NEXT)
+                            ievent.update_ha_state()
+                            self.nextevents[guid] = ievent
+                            self.nextevents[guid].keep_alive(starttime,
+                                                             endtime,
+                                                             duration,
+                                                             title,
+                                                             tzone,
+                                                             location)
+
+            except PyiCloudFailedLoginException as error:
+                _LOGGER.error('Error logging into iCloud Service: %s',
+                              error)
+
+    @property
+    def state(self):
+        """Return the state of the icloud account."""
+        return self.api is not None
+
+    @property
+    def state_attributes(self):
+        """Return the attributes of the icloud account."""
+        if self.getevents:
+            return {
+                ATTR_ACCOUNTNAME: self.accountname,
+                'current events': self._currentevents,
+                'next events': self._nextevents
+            }
+        else:
+            return {
+                ATTR_ACCOUNTNAME: self.accountname
+            }
+
+    @property
+    def icon(self):
+        """Return the icon to use for device if any."""
+        return 'mdi:account'
+
+    @staticmethod
+    def get_key(item):
+        """Sort key of events."""
+        return item.get('startDate')
+
+    def keep_alive(self):
+        """Keep the api alive."""
+        # pylint: disable=too-many-locals,too-many-branches,too-many-statements
+        if self.api is None:
+            from pyicloud import PyiCloudService
+            from pyicloud.exceptions import PyiCloudFailedLoginException
+            try:
+                # Attempt the login to iCloud
+                self.api = PyiCloudService(self.username,
+                                           self.password,
+                                           cookie_directory=self.cookiedir,
+                                           verify=True)
+            except PyiCloudFailedLoginException as error:
+                _LOGGER.error('Error logging into iCloud Service: %s',
+                              error)
+
+        if self.api is not None:
+            self.api.authenticate()
+            for devicename in self.devices:
+                self.devices[devicename].keep_alive()
+            if self.getevents:
+                from_dt = dt_util.now()
+                to_dt = from_dt + timedelta(days=7)
+                events = self.api.calendar.events(from_dt, to_dt)
+                new_events = sorted(events, key=self.get_key)
+                starttime = None
```

Comment:
There is no need set these values to ```None```. That applies to several places in the file.